src/main/java/org/mskcc/cmo/messaging/impl/NATSGatewayImpl.java

```diff
@@ -1,83 +1,82 @@
-
-import com.google.gson.Gson;
-import io.nats.client.Connection;
-import io.nats.client.Dispatcher;
-import io.nats.client.Message;
-import io.nats.client.MessageHandler;
-import java.nio.charset.StandardCharsets;
-import java.time.Duration;
-import java.util.*;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.beans.factory.annotation.Value;
 import org.springframework.stereotype.Component;
+import com.google.gson.Gson;
+import io.nats.streaming.Message;
+import io.nats.streaming.MessageHandler;
+import io.nats.streaming.StreamingConnection;
+import io.nats.streaming.Subscription;
+import io.nats.streaming.SubscriptionOptions;
 
 @Component
 public class NATSGatewayImpl implements Gateway {
-    @Value("${nats.flush_duration:5}")
-    private Integer natsFlushDuration;
-
+	
+	@Value("${nats.flush_duration:5}")
+	private Integer natsFlushDuration;
+	
+	@Value("${nats.durable_name}")
+	private String durableName;
+    
     @Autowired
-    private Connection natsConnection;
+    private StreamingConnection stanConnection;
 
     private Gson gson;
-    private Map<String, Dispatcher> dispatchers;
+    private Map<String, Subscription> subscribers;
     private boolean initialized;
 
 
-    public NATSGatewayImpl(Connection natsConnection) {
-        this.natsConnection = natsConnection;
+    public NATSGatewayImpl(StreamingConnection natsConnection) {
+        this.stanConnection = stanConnection;
         this.gson = new Gson();
-        this.dispatchers = new HashMap<>();
+        this.subscribers = new HashMap<>();
         this.initialized = (natsConnection != null);
     }
 
     public NATSGatewayImpl() {
         this.gson = new Gson();
-        this.dispatchers = new HashMap<>();
-        this.initialized = (natsConnection != null);
+        this.subscribers = new HashMap<>();
+        this.initialized = (stanConnection != null);
     }
 
     @Override
     public void publish(String topic, Object message) throws Exception {
         if (!initialized) return;
         String msg = gson.toJson(message);
-        natsConnection.publish(topic, msg.getBytes(StandardCharsets.UTF_8));
-        natsConnection.flush(Duration.ofSeconds(natsFlushDuration));
+        stanConnection.publish(topic, msg.getBytes(StandardCharsets.UTF_8));
     }
 
     @Override
     public void subscribe(String topic, Class messageClass, MessageConsumer consumer) throws Exception {
-        if (!initialized) return;
-        if (!dispatchers.containsKey(topic)) {
-            Dispatcher d = natsConnection.createDispatcher(new MessageHandler() {
-                @Override
-                public void onMessage(Message msg) throws InterruptedException {
-                    String json = new String(msg.getData(), StandardCharsets.UTF_8);
+    	if (!initialized) return;
+        if (!subscribers.containsKey(topic)) {
+        	final CountDownLatch doneSignal = new CountDownLatch(10);
```

Comment:
If I understand this correctly, this means that after 10 received messages, this thread exits (I think this also means, the consumer thread blocks until this happens).  We should remove this latch.